File: web/dashboard/src/lib/drDrill.ts
import type { AuthSession } from "./auth";
import { serviceRequest } from "./serviceApi";

export interface DrillSchedule {
  id: string;
  name: string;
  cron_expr: string;
  drill_type: "key_restore" | "full_failover" | "partial_restore" | "backup_verify";
  scope: "all_keys" | "critical_keys" | "selected";
  key_ids?: string[];
  target_env: string;
  enabled: boolean;
  last_run_at?: string;
  next_run_at?: string;
  created_at: string;
}

export interface DrillRun {
  id: string;
  schedule_id?: string;
  schedule_name?: string;
  drill_type: string;
  status: "running" | "passed" | "failed" | "partial" | "aborted";
  started_at: string;
  completed_at?: string;
  rto_seconds?: number;
  rpo_seconds?: number;
  total_keys: number;
  restored_keys: number;
  failed_keys: number;
  steps: DrillStep[];
  triggered_by: string;
  report_url?: string;
}

export interface DrillStep {
  name: string;
  status: "pending" | "running" | "passed" | "failed" | "skipped";
  started_at?: string;
  completed_at?: string;
  duration_ms?: number;
  message?: string;
}

export interface DrillMetrics {
  total_drills: number;
  pass_rate: number;
  avg_rto_seconds: number;
  last_30d_drills: number;
  last_failure_at?: string;
  trend: { date: string; rto: number; status: string }[];
}

export async function listSchedules(session: AuthSession): Promise<DrillSchedule[]> {
  const res = await serviceRequest<any>(session, "keycore", "/dr-drill/schedules");
  return res.data ?? [];
}

export async function createSchedule(session: AuthSession, data: Partial<DrillSchedule>): Promise<DrillSchedule> {
  return serviceRequest<DrillSchedule>(session, "keycore", "/dr-drill/schedules", { method: "POST", body: JSON.stringify(data) });
}

export async function deleteSchedule(session: AuthSession, id: string): Promise<void> {
  return serviceRequest<void>(session, "keycore", `/dr-drill/schedules/${id}`, { method: "DELETE" });
}

export async function triggerDrill(session: AuthSession, scheduleId?: string, type?: string): Promise<DrillRun> {
  const res = await serviceRequest<any>(session, "keycore", "/dr-drill/trigger", {
    method: "POST",
    body: JSON.stringify({ schedule_id: scheduleId, drill_type: type })
  });
  return res.data ?? res;
}

export async function listRuns(session: AuthSession): Promise<DrillRun[]> {
  const res = await serviceRequest<any>(session, "keycore", "/dr-drill/runs");
  return res.data ?? [];
}

export async function getRun(session: AuthSession, id: string): Promise<DrillRun> {
  const res = await serviceRequest<any>(session, "keycore", `/dr-drill/runs/${id}`);
  return res.data ?? res;
}

export async function getMetrics(session: AuthSession): Promise<DrillMetrics> {
  const res = await serviceRequest<any>(session, "keycore", "/dr-drill/metrics");
  return res.data ?? res;
}

File: services/keycore/drdrill.go
package main

import "time"

// DrillStepStatus represents the outcome of a single step in a DR drill.
type DrillStepStatus string

const (
	DrillStepStatusPending DrillStepStatus = "pending"
	DrillStepStatusRunning DrillStepStatus = "running"
	DrillStepStatusPassed  DrillStepStatus = "passed"
	DrillStepStatusFailed  DrillStepStatus = "failed"
	DrillStepStatusSkipped DrillStepStatus = "skipped"
)

// DrillStep records the outcome of one discrete step within a drill run.
type DrillStep struct {
	Name        string          `json:"name"`
	Status      DrillStepStatus `json:"status"`
	DurationMs  int64           `json:"duration_ms"`
	Detail      string          `json:"detail,omitempty"`
	ErrorMsg    string          `json:"error,omitempty"`
}

// DrillSchedule defines a recurring DR drill.
type DrillSchedule struct {
	ID         string     `json:"id"`
	TenantID   string     `json:"tenant_id"`
	Name       string     `json:"name"`
	CronExpr   string     `json:"cron_expr"`
	DrillType  string     `json:"drill_type"`
	Scope      string     `json:"scope"`
	TargetEnv  string     `json:"target_env"`
	Enabled    bool       `json:"enabled"`
	LastRunAt  *time.Time `json:"last_run_at,omitempty"`
	NextRunAt  *time.Time `json:"next_run_at,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
}

// DrillRun is a single execution of a DR drill, either triggered manually or
// by a schedule.
type DrillRun struct {
	ID           string      `json:"id"`
	TenantID     string      `json:"tenant_id"`
	ScheduleID   string      `json:"schedule_id,omitempty"`
	ScheduleName string      `json:"schedule_name,omitempty"`
	DrillType    string      `json:"drill_type"`
	Status       string      `json:"status"`
	StartedAt    time.Time   `json:"started_at"`
	CompletedAt  *time.Time  `json:"completed_at,omitempty"`
	RTOSeconds   int         `json:"rto_seconds,omitempty"`
	RPOSeconds   int         `json:"rpo_seconds,omitempty"`
	TotalKeys    int         `json:"total_keys"`
	RestoredKeys int         `json:"restored_keys"`
	FailedKeys   int         `json:"failed_keys"`
	Steps        []DrillStep `json:"steps"`
	TriggeredBy  string      `json:"triggered_by"`
}

// DrillMetrics aggregates statistics across all drill runs for a tenant.
type DrillMetrics struct {
	TotalRuns        int     `json:"total_runs"`
	SuccessfulRuns   int     `json:"successful_runs"`
	FailedRuns       int     `json:"failed_runs"`
	AvgRTOSeconds    float64 `json:"avg_rto_seconds"`
	AvgRPOSeconds    float64 `json:"avg_rpo_seconds"`
	AvgKeyRestoreRate float64 `json:"avg_key_restore_rate"` // percentage
	LastRunStatus    string  `json:"last_run_status"`
	LastRunAt        *time.Time `json:"last_run_at,omitempty"`
}
